// @bitrix24/b24jssdk/dist/cjs/frame/message/commands.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

var MessageCommands = /* @__PURE__ */ ((MessageCommands2) => {
  MessageCommands2["getInitData"] = "getInitData";
  MessageCommands2["setInstallFinish"] = "setInstallFinish";
  MessageCommands2["setInstall"] = "setInstall";
  MessageCommands2["refreshAuth"] = "refreshAuth";
  MessageCommands2["setAppOption"] = "setAppOption";
  MessageCommands2["setUserOption"] = "setUserOption";
  MessageCommands2["resizeWindow"] = "resizeWindow";
  MessageCommands2["reloadWindow"] = "reloadWindow";
  MessageCommands2["setTitle"] = "setTitle";
  MessageCommands2["setScroll"] = "setScroll";
  MessageCommands2["openApplication"] = "openApplication";
  MessageCommands2["closeApplication"] = "closeApplication";
  MessageCommands2["openPath"] = "openPath";
  MessageCommands2["imCallTo"] = "imCallTo";
  MessageCommands2["imPhoneTo"] = "imPhoneTo";
  MessageCommands2["imOpenMessenger"] = "imOpenMessenger";
  MessageCommands2["imOpenHistory"] = "imOpenHistory";
  MessageCommands2["selectUser"] = "selectUser";
  MessageCommands2["selectAccess"] = "selectAccess";
  MessageCommands2["selectCRM"] = "selectCRM";
  MessageCommands2["showAppForm"] = "showAppForm";
  MessageCommands2["getInterface"] = "getInterface";
  MessageCommands2["placementBindEvent"] = "placementBindEvent";
  return MessageCommands2;
})(MessageCommands || {});

exports.MessageCommands = MessageCommands;
