// @bitrix24/b24jssdk/dist/cjs/helper/abstract-helper.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const loggerFactory = require('../logger/logger-factory.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class UnhandledMatchError extends Error {
  static {
    __name(this, "UnhandledMatchError");
  }
  constructor(value, ...args) {
    super(...args);
    this.name = "UnhandledMatchError";
    this.message = `Unhandled match value of type ${value}`;
    this.stack = `${new Error("for stack").stack}`;
  }
}
class AbstractHelper {
  static {
    __name(this, "AbstractHelper");
  }
  _b24;
  _data = null;
  _logger;
  // region Init ////
  constructor(b24) {
    this._b24 = b24;
    this._logger = loggerFactory.LoggerFactory.createNullLogger();
  }
  setLogger(logger) {
    this._logger = logger;
  }
  getLogger() {
    return this._logger;
  }
  // endregion ////
  /**
   * Initializes the data received
   */
  async initData(_data) {
    return Promise.reject(new Error("Rewrite this function"));
  }
}

exports.AbstractHelper = AbstractHelper;
exports.UnhandledMatchError = UnhandledMatchError;
