// @bitrix24/b24jssdk/dist/cjs/logger/handler/consola-adapter.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const logger = require('../../types/logger.cjs');
const abstractHandler = require('./abstract-handler.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class ConsolaAdapter extends abstractHandler.AbstractHandler {
  static {
    __name(this, "ConsolaAdapter");
  }
  consolaInstance;
  constructor(level = logger.LogLevel.DEBUG, options) {
    const opts = {
      bubble: true,
      ...options
    };
    super(level, opts.bubble);
    this.consolaInstance = opts.consolaInstance;
  }
  setFormatter(_formatter) {
  }
  getFormatter() {
    return null;
  }
  async handle(record) {
    const message = `[${record.channel}] ${record.levelName}: ${record.message}`;
    const args = { ...record.context, ...record.extra, timestamp: record.timestamp };
    switch (record.level) {
      case logger.LogLevel.DEBUG:
        this.consolaInstance.log(message, args);
        break;
      case logger.LogLevel.INFO:
        this.consolaInstance.info(message, args);
        break;
      case logger.LogLevel.NOTICE:
        this.consolaInstance.success(message, args);
        break;
      case logger.LogLevel.WARNING:
        this.consolaInstance.warn(message, args);
        break;
      case logger.LogLevel.ERROR:
      case logger.LogLevel.CRITICAL:
      case logger.LogLevel.ALERT:
      case logger.LogLevel.EMERGENCY:
        this.consolaInstance.error(message, args);
        break;
      default:
        this.consolaInstance.log(message, args);
    }
    return true;
  }
}

exports.ConsolaAdapter = ConsolaAdapter;
