// @bitrix24/b24jssdk/dist/cjs/logger/handler/memory-handler.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const logger = require('../../types/logger.cjs');
const abstractHandler = require('./abstract-handler.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class MemoryHandler extends abstractHandler.AbstractHandler {
  static {
    __name(this, "MemoryHandler");
  }
  records = [];
  limit;
  constructor(level = logger.LogLevel.DEBUG, options) {
    const opts = {
      bubble: true,
      limit: 1e3,
      ...options
    };
    super(level, opts.bubble);
    this.limit = opts.limit;
  }
  /**
   * @inheritDoc
   */
  async handle(record) {
    this.records.push(record);
    if (this.records.length > this.limit) {
      this.records.shift();
    }
    return true;
  }
  getRecords() {
    return [...this.records];
  }
  clear() {
    this.records = [];
  }
}

exports.MemoryHandler = MemoryHandler;
