// @bitrix24/b24jssdk/dist/cjs/logger/handler/winston-adapter.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const logger = require('../../types/logger.cjs');
const abstractHandler = require('./abstract-handler.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class WinstonAdapter extends abstractHandler.AbstractHandler {
  static {
    __name(this, "WinstonAdapter");
  }
  winstonLogger;
  constructor(level = logger.LogLevel.DEBUG, options) {
    const opts = {
      bubble: true,
      ...options
    };
    super(level, opts.bubble);
    this.winstonLogger = opts.winstonLogger;
  }
  setFormatter(_formatter) {
  }
  getFormatter() {
    return null;
  }
  async handle(record) {
    const levelMap = {
      DEBUG: "debug",
      INFO: "info",
      NOTICE: "notice",
      WARNING: "warn",
      ERROR: "error",
      CRITICAL: "error",
      ALERT: "error",
      EMERGENCY: "error"
    };
    const winstonLevel = levelMap[record.levelName] || levelMap.INFO;
    this.winstonLogger.log({
      level: winstonLevel,
      channel: record.channel,
      message: record.message,
      context: record.context,
      extra: record.extra,
      timestamp: record.timestamp
    });
    return true;
  }
}

exports.WinstonAdapter = WinstonAdapter;
