// @bitrix24/b24jssdk/dist/cjs/logger/logger-factory.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const logger = require('../types/logger.cjs');
const logger$1 = require('./logger.cjs');
const nullLogger = require('./null-logger.cjs');
const consoleV2Handler = require('./handler/console-v2-handler.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class LoggerFactory {
  static {
    __name(this, "LoggerFactory");
  }
  static createNullLogger() {
    return nullLogger.NullLogger.create();
  }
  static createForBrowser(channel, isDevMode = false) {
    if (isDevMode) {
      return LoggerFactory.createForBrowserDevelopment(channel);
    }
    return LoggerFactory.createForBrowserProduction(channel);
  }
  static createForBrowserDevelopment(channel, level = logger.LogLevel.DEBUG) {
    const logger = new logger$1.Logger(channel);
    const handler = new consoleV2Handler.ConsoleV2Handler(level);
    logger.pushHandler(handler);
    return logger;
  }
  static createForBrowserProduction(channel, level = logger.LogLevel.ERROR) {
    const logger = new logger$1.Logger(channel);
    const handler = new consoleV2Handler.ConsoleV2Handler(level);
    logger.pushHandler(handler);
    return logger;
  }
  static async forcedLog(logger, action, message, context) {
    if (typeof globalThis !== "undefined" && "vitest" in globalThis) {
      return;
    }
    if (logger instanceof nullLogger.NullLogger) {
      switch (action) {
        case "debug":
          console.log(message, context);
          return;
        case "info":
        case "notice":
          console.info(message, context);
          return;
        case "warning":
          console.warn(message, context);
          return;
        default:
          console.error(message, context);
          return;
      }
    }
    return logger[action](message, context);
  }
}

exports.LoggerFactory = LoggerFactory;
