// @bitrix24/b24jssdk/dist/esm/core/actions/v2/manager-v2.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { CallV2 } from './call.mjs';
import { CallListV2 } from './call-list.mjs';
import { FetchListV2 } from './fetch-list.mjs';
import { BatchV2 } from './batch.mjs';
import { BatchByChunkV2 } from './batch-by-chunk.mjs';
import { LoggerFactory } from '../../../logger/logger-factory.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const callName = /* @__PURE__ */ Symbol("call_V2");
const callListName = /* @__PURE__ */ Symbol("callList_V2");
const fetchListName = /* @__PURE__ */ Symbol("fetchList_V2");
const batchName = /* @__PURE__ */ Symbol("batch_V2");
const batchByChunkName = /* @__PURE__ */ Symbol("batchByChunk_V2");
class ActionsManagerV2 {
  static {
    __name(this, "ActionsManagerV2");
  }
  _b24;
  _logger;
  _mapActions;
  constructor(b24) {
    this._b24 = b24;
    this._logger = LoggerFactory.createNullLogger();
    this._mapActions = /* @__PURE__ */ new Map();
  }
  setLogger(logger) {
    this._logger = logger;
  }
  getLogger() {
    return this._logger;
  }
  get call() {
    if (!this._mapActions.has(callName)) {
      this._mapActions.set(callName, new CallV2(this._b24, this._logger));
    }
    return this._mapActions.get(callName);
  }
  get callList() {
    if (!this._mapActions.has(callListName)) {
      this._mapActions.set(callListName, new CallListV2(this._b24, this._logger));
    }
    return this._mapActions.get(callListName);
  }
  get fetchList() {
    if (!this._mapActions.has(fetchListName)) {
      this._mapActions.set(fetchListName, new FetchListV2(this._b24, this._logger));
    }
    return this._mapActions.get(fetchListName);
  }
  get batch() {
    if (!this._mapActions.has(batchName)) {
      this._mapActions.set(batchName, new BatchV2(this._b24, this._logger));
    }
    return this._mapActions.get(batchName);
  }
  get batchByChunk() {
    if (!this._mapActions.has(batchByChunkName)) {
      this._mapActions.set(batchByChunkName, new BatchByChunkV2(this._b24, this._logger));
    }
    return this._mapActions.get(batchByChunkName);
  }
}

export { ActionsManagerV2 };
