// @bitrix24/b24jssdk/dist/cjs/core/language/list.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

var B24LangList = /* @__PURE__ */ ((B24LangList2) => {
  B24LangList2["ru"] = "ru";
  B24LangList2["id"] = "id";
  B24LangList2["ms"] = "ms";
  B24LangList2["de"] = "de";
  B24LangList2["en"] = "en";
  B24LangList2["la"] = "la";
  B24LangList2["fr"] = "fr";
  B24LangList2["in"] = "in";
  B24LangList2["it"] = "it";
  B24LangList2["pl"] = "pl";
  B24LangList2["br"] = "br";
  B24LangList2["vn"] = "vn";
  B24LangList2["tr"] = "tr";
  B24LangList2["kz"] = "kz";
  B24LangList2["ua"] = "ua";
  B24LangList2["ar"] = "ar";
  B24LangList2["th"] = "th";
  B24LangList2["sc"] = "sc";
  B24LangList2["tc"] = "tc";
  B24LangList2["ja"] = "ja";
  return B24LangList2;
})(B24LangList || {});
const B24LocaleMap = {
  ["ru" /* ru */]: "ru-RU",
  ["id" /* id */]: "id-ID",
  ["ms" /* ms */]: "ms-MY",
  ["de" /* de */]: "de-DE",
  ["en" /* en */]: "en-EN",
  ["la" /* la */]: "es-ES",
  ["fr" /* fr */]: "fr-FR",
  ["in" /* in */]: "hi-IN",
  ["it" /* it */]: "it-IT",
  ["pl" /* pl */]: "pl-PL",
  ["br" /* br */]: "pt-BR",
  ["vn" /* vn */]: "vi-VN",
  ["tr" /* tr */]: "tr-TR",
  ["kz" /* kz */]: "kk",
  ["ua" /* ua */]: "uk-UA",
  ["ar" /* ar */]: "ar-SA",
  ["th" /* th */]: "th-TH",
  ["sc" /* sc */]: "zh-CN",
  ["tc" /* tc */]: "zh-TW",
  ["ja" /* ja */]: "ja-JP"
};

exports.B24LangList = B24LangList;
exports.B24LocaleMap = B24LocaleMap;
