// @bitrix24/b24jssdk/dist/esm/logger/abstract-logger.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { LogLevel } from '../types/logger.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class AbstractLogger {
  static {
    __name(this, "AbstractLogger");
  }
  /**
   * @inheritDoc
   */
  async debug(message, context) {
    return this.log(LogLevel.DEBUG, message, context);
  }
  /**
   * @inheritDoc
   */
  async info(message, context) {
    return this.log(LogLevel.INFO, message, context);
  }
  /**
   * @inheritDoc
   */
  async notice(message, context) {
    return this.log(LogLevel.NOTICE, message, context);
  }
  /**
   * @inheritDoc
   */
  async warning(message, context) {
    return this.log(LogLevel.WARNING, message, context);
  }
  /**
   * @inheritDoc
   */
  async error(message, context) {
    return this.log(LogLevel.ERROR, message, context);
  }
  /**
   * @inheritDoc
   */
  async critical(message, context) {
    return this.log(LogLevel.CRITICAL, message, context);
  }
  /**
   * @inheritDoc
   */
  async alert(message, context) {
    return this.log(LogLevel.ALERT, message, context);
  }
  /**
   * @inheritDoc
   */
  async emergency(message, context) {
    return this.log(LogLevel.EMERGENCY, message, context);
  }
  // endregion ////
}

export { AbstractLogger };
