// @bitrix24/b24jssdk/dist/esm/logger/browser.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { LoggerFactory } from './logger-factory.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const deprecateMessage = "@deprecate: use Logger. https://bitrix24.github.io/b24jssdk/docs/working-with-the-rest-api/logger/";
var LoggerType = /* @__PURE__ */ ((LoggerType2) => {
  LoggerType2["desktop"] = "desktop";
  LoggerType2["log"] = "log";
  LoggerType2["info"] = "info";
  LoggerType2["warn"] = "warn";
  LoggerType2["error"] = "error";
  LoggerType2["trace"] = "trace";
  return LoggerType2;
})(LoggerType || {});
class LoggerBrowser {
  static {
    __name(this, "LoggerBrowser");
  }
  #logger;
  /**
   * Create a LoggerBrowser instance
   *
   * @deprecated This method is deprecated and will be removed in version `2.0.0`
   *
   * @removed 2.0.0
   */
  static build(title, isDevelopment = false) {
    return new LoggerBrowser(title, isDevelopment);
  }
  constructor(title, isDevelopment = false) {
    console.warn(deprecateMessage);
    if (isDevelopment) {
      this.#logger = LoggerFactory.createForBrowserDevelopment(title);
    } else {
      this.#logger = LoggerFactory.createForBrowserProduction(title);
    }
  }
  // region Config ////
  /**
   * Set config
   *
   * @deprecated This method is deprecated and will be removed in version `2.0.0`
   *
   * @removed 2.0.0
   */
  setConfig(_types) {
    console.warn(deprecateMessage);
  }
  /**
   * Set enable
   *
   * @deprecated This method is deprecated and will be removed in version `2.0.0`
   *
   * @removed 2.0.0
   */
  enable(_type) {
    console.warn(deprecateMessage);
    return true;
  }
  /**
   * Set disable
   *
   * @deprecated This method is deprecated and will be removed in version `2.0.0`
   *
   * @removed 2.0.0
   */
  disable(_type) {
    console.warn(deprecateMessage);
    return true;
  }
  /**
   * Test is enable
   *
   * @deprecated This method is deprecated and will be removed in version `2.0.0`
   *
   * @removed 2.0.0
   */
  isEnabled(_type) {
    console.warn(deprecateMessage);
    return false;
  }
  // endregion ////
  // region Functions ////
  async desktop(...params) {
    console.warn(deprecateMessage);
    const context = {
      needDesktop: true,
      params: { ...params }
    };
    return this.#logger.debug("desktop", context);
  }
  async log(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.debug("log", context);
  }
  async info(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.info("info", context);
  }
  async warn(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.warning("warn", context);
  }
  async error(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.error("error", context);
  }
  async trace(...params) {
    console.warn(deprecateMessage);
    const context = {
      needTrace: true,
      params: { ...params }
    };
    return this.#logger.debug("trace", context);
  }
  async debug(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.debug("debug", context);
  }
  async notice(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.info("notice", context);
  }
  async warning(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.warning("warning", context);
  }
  async critical(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.critical("critical", context);
  }
  async alert(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.alert("alert", context);
  }
  async emergency(...params) {
    console.warn(deprecateMessage);
    const context = { params: { ...params } };
    return this.#logger.emergency("alert", context);
  }
  // endregion ////
}

export { LoggerBrowser, LoggerType };
