// @bitrix24/b24jssdk/dist/esm/types/pull.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
var ConnectionType = /* @__PURE__ */ ((ConnectionType2) => {
  ConnectionType2["Undefined"] = "undefined";
  ConnectionType2["WebSocket"] = "webSocket";
  ConnectionType2["LongPolling"] = "longPolling";
  return ConnectionType2;
})(ConnectionType || {});
var LsKeys = /* @__PURE__ */ ((LsKeys2) => {
  LsKeys2["PullConfig"] = "bx-pull-config";
  LsKeys2["WebsocketBlocked"] = "bx-pull-websocket-blocked";
  LsKeys2["LongPollingBlocked"] = "bx-pull-longpolling-blocked";
  LsKeys2["LoggingEnabled"] = "bx-pull-logging-enabled";
  return LsKeys2;
})(LsKeys || {});
var PullStatus = /* @__PURE__ */ ((PullStatus2) => {
  PullStatus2["Online"] = "online";
  PullStatus2["Offline"] = "offline";
  PullStatus2["Connecting"] = "connect";
  return PullStatus2;
})(PullStatus || {});
var SenderType = /* @__PURE__ */ ((SenderType2) => {
  SenderType2[SenderType2["Unknown"] = 0] = "Unknown";
  SenderType2[SenderType2["Client"] = 1] = "Client";
  SenderType2[SenderType2["Backend"] = 2] = "Backend";
  return SenderType2;
})(SenderType || {});
var SubscriptionType = /* @__PURE__ */ ((SubscriptionType2) => {
  SubscriptionType2["Server"] = "server";
  SubscriptionType2["Client"] = "client";
  SubscriptionType2["Online"] = "online";
  SubscriptionType2["Status"] = "status";
  SubscriptionType2["Revision"] = "revision";
  return SubscriptionType2;
})(SubscriptionType || {});
var CloseReasons = /* @__PURE__ */ ((CloseReasons2) => {
  CloseReasons2[CloseReasons2["NORMAL_CLOSURE"] = 1e3] = "NORMAL_CLOSURE";
  CloseReasons2[CloseReasons2["SERVER_DIE"] = 1001] = "SERVER_DIE";
  CloseReasons2[CloseReasons2["CONFIG_REPLACED"] = 3e3] = "CONFIG_REPLACED";
  CloseReasons2[CloseReasons2["CHANNEL_EXPIRED"] = 3001] = "CHANNEL_EXPIRED";
  CloseReasons2[CloseReasons2["SERVER_RESTARTED"] = 3002] = "SERVER_RESTARTED";
  CloseReasons2[CloseReasons2["CONFIG_EXPIRED"] = 3003] = "CONFIG_EXPIRED";
  CloseReasons2[CloseReasons2["MANUAL"] = 3004] = "MANUAL";
  CloseReasons2[CloseReasons2["STUCK"] = 3005] = "STUCK";
  CloseReasons2[CloseReasons2["WRONG_CHANNEL_ID"] = 4010] = "WRONG_CHANNEL_ID";
  return CloseReasons2;
})(CloseReasons || {});
var SystemCommands = /* @__PURE__ */ ((SystemCommands2) => {
  SystemCommands2["CHANNEL_EXPIRE"] = "CHANNEL_EXPIRE";
  SystemCommands2["CONFIG_EXPIRE"] = "CONFIG_EXPIRE";
  SystemCommands2["SERVER_RESTART"] = "SERVER_RESTART";
  return SystemCommands2;
})(SystemCommands || {});
var ServerMode = /* @__PURE__ */ ((ServerMode2) => {
  ServerMode2["Shared"] = "shared";
  ServerMode2["Personal"] = "personal";
  return ServerMode2;
})(ServerMode || {});
const ListRpcError = {
  Parse: { code: -32700, message: "Parse error" },
  InvalidRequest: { code: -32600, message: "Invalid Request" },
  MethodNotFound: { code: -32601, message: "Method not found" },
  InvalidParams: { code: -32602, message: "Invalid params" },
  Internal: { code: -32603, message: "Internal error" }
};
var RpcMethod = /* @__PURE__ */ ((RpcMethod2) => {
  RpcMethod2["Publish"] = "publish";
  RpcMethod2["GetUsersLastSeen"] = "getUsersLastSeen";
  RpcMethod2["Ping"] = "ping";
  RpcMethod2["ListChannels"] = "listChannels";
  RpcMethod2["SubscribeStatusChange"] = "subscribeStatusChange";
  RpcMethod2["UnsubscribeStatusChange"] = "unsubscribeStatusChange";
  return RpcMethod2;
})(RpcMethod || {});

export { CloseReasons, ConnectionType, ListRpcError, LsKeys, PullStatus, RpcMethod, SenderType, ServerMode, SubscriptionType, SystemCommands };
