// @bitrix24/b24jssdk/dist/esm/core/actions/v3/manager-v3.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { CallV3 } from './call.mjs';
import { CallListV3 } from './call-list.mjs';
import { FetchListV3 } from './fetch-list.mjs';
import { CallTailV3 } from './call-tail.mjs';
import { FetchTailV3 } from './fetch-tail.mjs';
import { AggregateV3 } from './aggregate.mjs';
import { BatchV3 } from './batch.mjs';
import { BatchByChunkV3 } from './batch-by-chunk.mjs';
import { LoggerFactory } from '../../../logger/logger-factory.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const callName = /* @__PURE__ */ Symbol("call_V3");
const callListName = /* @__PURE__ */ Symbol("callList_V3");
const fetchListName = /* @__PURE__ */ Symbol("fetchList_V3");
const callTailName = /* @__PURE__ */ Symbol("callTail_V3");
const fetchTailName = /* @__PURE__ */ Symbol("fetchTail_V3");
const aggregateName = /* @__PURE__ */ Symbol("aggregate_V3");
const batchName = /* @__PURE__ */ Symbol("batch_V3");
const batchByChunkName = /* @__PURE__ */ Symbol("batchByChunk_V3");
class ActionsManagerV3 {
  static {
    __name(this, "ActionsManagerV3");
  }
  _b24;
  _logger;
  _mapActions;
  constructor(b24) {
    this._b24 = b24;
    this._logger = LoggerFactory.createNullLogger();
    this._mapActions = /* @__PURE__ */ new Map();
  }
  setLogger(logger) {
    this._logger = logger;
  }
  getLogger() {
    return this._logger;
  }
  get call() {
    if (!this._mapActions.has(callName)) {
      this._mapActions.set(callName, new CallV3(this._b24, this._logger));
    }
    return this._mapActions.get(callName);
  }
  get callList() {
    if (!this._mapActions.has(callListName)) {
      this._mapActions.set(callListName, new CallListV3(this._b24, this._logger));
    }
    return this._mapActions.get(callListName);
  }
  get fetchList() {
    if (!this._mapActions.has(fetchListName)) {
      this._mapActions.set(fetchListName, new FetchListV3(this._b24, this._logger));
    }
    return this._mapActions.get(fetchListName);
  }
  get callTail() {
    if (!this._mapActions.has(callTailName)) {
      this._mapActions.set(callTailName, new CallTailV3(this._b24, this._logger));
    }
    return this._mapActions.get(callTailName);
  }
  get fetchTail() {
    if (!this._mapActions.has(fetchTailName)) {
      this._mapActions.set(fetchTailName, new FetchTailV3(this._b24, this._logger));
    }
    return this._mapActions.get(fetchTailName);
  }
  get aggregate() {
    if (!this._mapActions.has(aggregateName)) {
      this._mapActions.set(aggregateName, new AggregateV3(this._b24, this._logger));
    }
    return this._mapActions.get(aggregateName);
  }
  get batch() {
    if (!this._mapActions.has(batchName)) {
      this._mapActions.set(batchName, new BatchV3(this._b24, this._logger));
    }
    return this._mapActions.get(batchName);
  }
  get batchByChunk() {
    if (!this._mapActions.has(batchByChunkName)) {
      this._mapActions.set(batchByChunkName, new BatchByChunkV3(this._b24, this._logger));
    }
    return this._mapActions.get(batchByChunkName);
  }
}

export { ActionsManagerV3 };
