// @bitrix24/b24jssdk/dist/cjs/types/b24-helper.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

var LoadDataType = /* @__PURE__ */ ((LoadDataType2) => {
  LoadDataType2["App"] = "app";
  LoadDataType2["Profile"] = "profile";
  LoadDataType2["Currency"] = "currency";
  LoadDataType2["AppOptions"] = "appOptions";
  LoadDataType2["UserOptions"] = "userOptions";
  return LoadDataType2;
})(LoadDataType || {});
const EnumAppStatus = {
  // free ////
  Free: "F",
  // demo version ////
  Demo: "D",
  // trial version (limited time) ////
  Trial: "T",
  // paid application ////
  Paid: "P",
  // local application ////
  Local: "L",
  // subscription application ////
  Subscription: "S"
};
const StatusDescriptions = {
  [EnumAppStatus.Free]: "Free",
  [EnumAppStatus.Demo]: "Demo",
  [EnumAppStatus.Trial]: "Trial",
  [EnumAppStatus.Paid]: "Paid",
  [EnumAppStatus.Local]: "Local",
  [EnumAppStatus.Subscription]: "Subscription"
};
const TypeSpecificUrl = {
  MainSettings: "MainSettings",
  UfList: "UfList",
  UfPage: "UfPage"
};
var TypeOption = /* @__PURE__ */ ((TypeOption2) => {
  TypeOption2["NotSet"] = "notSet";
  TypeOption2["JsonArray"] = "jsonArray";
  TypeOption2["JsonObject"] = "jsonObject";
  TypeOption2["FloatVal"] = "float";
  TypeOption2["IntegerVal"] = "integer";
  TypeOption2["BoolYN"] = "boolYN";
  TypeOption2["StringVal"] = "string";
  return TypeOption2;
})(TypeOption || {});

exports.EnumAppStatus = EnumAppStatus;
exports.LoadDataType = LoadDataType;
exports.StatusDescriptions = StatusDescriptions;
exports.TypeOption = TypeOption;
exports.TypeSpecificUrl = TypeSpecificUrl;
