// @bitrix24/b24jssdk/dist/esm/types/crm/entity-type.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var EnumCrmEntityType = /* @__PURE__ */ ((EnumCrmEntityType2) => {
  EnumCrmEntityType2["undefined"] = "UNDEFINED";
  EnumCrmEntityType2["lead"] = "CRM_LEAD";
  EnumCrmEntityType2["deal"] = "CRM_DEAL";
  EnumCrmEntityType2["contact"] = "CRM_CONTACT";
  EnumCrmEntityType2["company"] = "CRM_COMPANY";
  EnumCrmEntityType2["oldInvoice"] = "CRM_INVOICE";
  EnumCrmEntityType2["invoice"] = "CRM_SMART_INVOICE";
  EnumCrmEntityType2["quote"] = "CRM_QUOTE";
  EnumCrmEntityType2["requisite"] = "CRM_REQUISITE";
  EnumCrmEntityType2["order"] = "ORDER";
  return EnumCrmEntityType2;
})(EnumCrmEntityType || {});
var EnumCrmEntityTypeId = /* @__PURE__ */ ((EnumCrmEntityTypeId2) => {
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["undefined"] = 0] = "undefined";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["lead"] = 1] = "lead";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["deal"] = 2] = "deal";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["contact"] = 3] = "contact";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["company"] = 4] = "company";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["oldInvoice"] = 5] = "oldInvoice";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["invoice"] = 31] = "invoice";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["quote"] = 7] = "quote";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["requisite"] = 8] = "requisite";
  EnumCrmEntityTypeId2[EnumCrmEntityTypeId2["order"] = 14] = "order";
  return EnumCrmEntityTypeId2;
})(EnumCrmEntityTypeId || {});
var EnumCrmEntityTypeShort = /* @__PURE__ */ ((EnumCrmEntityTypeShort2) => {
  EnumCrmEntityTypeShort2["undefined"] = "?";
  EnumCrmEntityTypeShort2["lead"] = "L";
  EnumCrmEntityTypeShort2["deal"] = "D";
  EnumCrmEntityTypeShort2["contact"] = "C";
  EnumCrmEntityTypeShort2["company"] = "CO";
  EnumCrmEntityTypeShort2["oldInvoice"] = "I";
  EnumCrmEntityTypeShort2["invoice"] = "SI";
  EnumCrmEntityTypeShort2["quote"] = "Q";
  EnumCrmEntityTypeShort2["requisite"] = "RQ";
  EnumCrmEntityTypeShort2["order"] = "O";
  return EnumCrmEntityTypeShort2;
})(EnumCrmEntityTypeShort || {});
function getEnumCrmEntityTypeShort(id) {
  const key = EnumCrmEntityTypeId[id];
  return EnumCrmEntityTypeShort[key] || "?" /* undefined */;
}
__name(getEnumCrmEntityTypeShort, "getEnumCrmEntityTypeShort");

export { EnumCrmEntityType, EnumCrmEntityTypeId, EnumCrmEntityTypeShort, getEnumCrmEntityTypeShort };
