// @bitrix24/b24jssdk/dist/esm/types/catalog/index.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
var CatalogProductType = /* @__PURE__ */ ((CatalogProductType2) => {
  CatalogProductType2[CatalogProductType2["undefined"] = 0] = "undefined";
  CatalogProductType2[CatalogProductType2["product"] = 1] = "product";
  CatalogProductType2[CatalogProductType2["service"] = 7] = "service";
  CatalogProductType2[CatalogProductType2["sku"] = 3] = "sku";
  CatalogProductType2[CatalogProductType2["skuEmpty"] = 6] = "skuEmpty";
  CatalogProductType2[CatalogProductType2["offer"] = 4] = "offer";
  CatalogProductType2[CatalogProductType2["offerEmpty"] = 5] = "offerEmpty";
  return CatalogProductType2;
})(CatalogProductType || {});
var CatalogProductImageType = /* @__PURE__ */ ((CatalogProductImageType2) => {
  CatalogProductImageType2["undefined"] = "UNDEFINED";
  CatalogProductImageType2["detail"] = "DETAIL_PICTURE";
  CatalogProductImageType2["preview"] = "PREVIEW_PICTURE";
  CatalogProductImageType2["morePhoto"] = "MORE_PHOTO";
  return CatalogProductImageType2;
})(CatalogProductImageType || {});
var CatalogRoundingRuleType = /* @__PURE__ */ ((CatalogRoundingRuleType2) => {
  CatalogRoundingRuleType2[CatalogRoundingRuleType2["undefined"] = 0] = "undefined";
  CatalogRoundingRuleType2[CatalogRoundingRuleType2["mathematical"] = 1] = "mathematical";
  CatalogRoundingRuleType2[CatalogRoundingRuleType2["roundingUp"] = 2] = "roundingUp";
  CatalogRoundingRuleType2[CatalogRoundingRuleType2["roundingDown"] = 4] = "roundingDown";
  return CatalogRoundingRuleType2;
})(CatalogRoundingRuleType || {});

export { CatalogProductImageType, CatalogProductType, CatalogRoundingRuleType };
