// @bitrix24/b24jssdk/dist/cjs/core/actions/v2/call-list.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const abstractAction = require('../abstract-action.cjs');
const result = require('../../result.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class CallListV2 extends abstractAction.AbstractAction {
  static {
    __name(this, "CallListV2");
  }
  /**
   * Fast data retrieval without counting the total number of records.
   *
   * @template T - The type of the elements of the returned array (default is `unknown`).
   *
   * @param {ActionCallListV2} options - parameters for executing the request.
   *     - `method: string` - The name of the REST API method that returns a list of data (for example: `crm.item.list`, `tasks.task.list`)
   *     - `params?: Omit<TypeCallParams, 'start' | 'order'>` - Request parameters, excluding the `start` and `order` parameters,
   *         since the method is designed to obtain all data in one call.
   *         Note: Use `filter`, `order`, and `select` to control the selection.
   *     - `idKey?: string` - The name of the id field as it appears in each RESPONSE item; its value
   *         drives the cursor. Default is 'ID' (uppercase). For methods that return a lowercase /
   *         camelCase id (for example `tasks.task.list` returns `id`), set `idKey: 'id'`.
   *     - `cursorIdKey?: string` - The field name used in the REQUEST for `order` and the `>` page
   *         filter. Defaults to `idKey`. Set it only when the sortable / filterable field name differs
   *         from the response field name — e.g. `tasks.task.list` sorts and filters by `ID` (uppercase)
   *         but returns `id` (lowercase): pass `idKey: 'id', cursorIdKey: 'ID'`.
   *     - `customKeyForResult?: string` - A custom key indicating that the response REST API will be
   *        grouped by this field.
   *        Example: `items` to group a list of CRM items.
   *    - `requestId?: string` - Unique request identifier for tracking. Used for query deduplication and debugging.
   *
   * @returns {Promise<Result<T[]>>} A promise that resolves to the result of an REST API call.
   *
   * @example
   * import { EnumCrmEntityTypeId, Text } from '@bitrix24/b24jssdk'
   *
   * interface CrmItem { id: number, title: string }
   * const sixMonthAgo = new Date()
   * sixMonthAgo.setMonth((new Date()).getMonth() - 6)
   * sixMonthAgo.setHours(0, 0, 0)
   * const response = await b24.actions.v2.callList.make<CrmItem>({
   *   method: 'crm.item.list',
   *   params: {
   *     entityTypeId:  EnumCrmEntityTypeId.company,
   *     filter: {
   *       '=%title': 'A%',
   *       '>=createdTime': Text.toB24Format(sixMonthAgo) // created at least 6 months ago
   *     },
   *     select: ['id', 'title']
   *   },
   *   idKey: 'id',
   *   customKeyForResult: 'items',
   *   requestId: 'list-123'
   * })
   * if (!response.isSuccess) {
   *   throw new Error(`Problem: ${response.getErrorMessages().join('; ')}`)
   * }
   * const list = response.getData()
   * console.log(`Result: ${list?.length}`) // Number of items received
   */
  async make(options) {
    const batchSize = 50;
    const result$1 = new result.Result();
    const idKey = options?.idKey ?? "ID";
    const cursorIdKey = options?.cursorIdKey ?? idKey;
    const customKeyForResult = options?.customKeyForResult ?? null;
    const params = options?.params ?? {};
    if ("order" in params && params["order"]) {
      this._logger.warning("callList.make: user-provided `order` parameter is ignored because cursor-based pagination requires ordering by cursorIdKey. Use `filter` to narrow results instead.");
    }
    const moreIdKey = `>${cursorIdKey}`;
    const { order: _ignoredOrder, ...restParams } = params;
    const requestParams = {
      ...restParams,
      order: { [cursorIdKey]: "ASC" },
      filter: { ...params["filter"] || {}, [moreIdKey]: 0 },
      start: -1
    };
    let allItems = [];
    let isContinue = true;
    do {
      const response = await this._b24.actions.v2.call.make({
        method: options.method,
        params: requestParams,
        requestId: options.requestId
      });
      if (!response.isSuccess) {
        this._logger.error("callFastListMethod", {
          method: options.method,
          requestId: options.requestId,
          messages: response.getErrorMessages()
        });
        for (const [index, error] of response.errors) {
          result$1.addError(error, index);
        }
        isContinue = false;
        break;
      }
      const responseData = response.getData();
      if (!responseData) {
        isContinue = false;
        break;
      }
      let resultData = [];
      if (null === customKeyForResult) {
        resultData = responseData.result;
      } else {
        resultData = responseData.result[customKeyForResult];
      }
      if (resultData.length === 0) {
        isContinue = false;
        break;
      }
      allItems = [...allItems, ...resultData];
      if (resultData.length < batchSize) {
        isContinue = false;
        break;
      }
      const lastItem = resultData[resultData.length - 1];
      const cursorValue = lastItem ? Number.parseInt(lastItem[idKey], 10) : Number.NaN;
      if (Number.isFinite(cursorValue)) {
        requestParams.filter[moreIdKey] = cursorValue;
      } else {
        this._logger.warning(`callList.make: pagination stops here \u2014 no numeric id could be read from the returned items via idKey "${idKey}". Make sure idKey matches the id field in the response; if the sortable field name differs from it, also set cursorIdKey (e.g. idKey: 'id', cursorIdKey: 'ID').`);
        isContinue = false;
        break;
      }
    } while (isContinue);
    return result$1.setData(allItems);
  }
}

exports.CallListV2 = CallListV2;
