// @bitrix24/b24jssdk/dist/cjs/core/interaction/batch/v2.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const abstractInteractionBatch = require('./abstract-interaction-batch.cjs');
const sdkError = require('../../sdk-error.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const MAX_BATCH_COMMANDS_V2 = 50;
class InteractionBatchV2 extends abstractInteractionBatch.AbstractInteractionBatch {
  static {
    __name(this, "InteractionBatchV2");
  }
  get maxSize() {
    return MAX_BATCH_COMMANDS_V2;
  }
  // region Response ////
  async prepareResponse(response) {
    if (!this.processingStrategy) {
      throw new sdkError.SdkError({
        code: "JSSDK_INTERACTION_BATCH_EMPTY_PROCESSING_STRATEGY",
        description: "ProcessingStrategy not set",
        status: 500
      });
    }
    const responseHelper = {
      requestId: response.getQuery().requestId,
      parallelDefaultValue: this.parallelDefaultValue,
      restrictionManager: this.restrictionManager,
      response
    };
    const results = await this.processingStrategy.prepareItems(this._commands, responseHelper);
    return this.processingStrategy.handleResults(this._commands, results, responseHelper);
  }
  // endregion ////
}

exports.InteractionBatchV2 = InteractionBatchV2;
exports.MAX_BATCH_COMMANDS_V2 = MAX_BATCH_COMMANDS_V2;
