// @bitrix24/b24jssdk/dist/cjs/index.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const common = require('./types/common.cjs');
const type = require('./tools/type.cjs');
const index$2 = require('./tools/index.cjs');
const text = require('./tools/text.cjs');
const browser = require('./tools/browser.cjs');
const b24 = require('./types/b24.cjs');
const index = require('./types/catalog/index.cjs');
const index$1 = require('./types/bizproc/index.cjs');
const b24Helper = require('./types/b24-helper.cjs');
const pull = require('./types/pull.cjs');
const list = require('./core/language/list.cjs');
const result = require('./core/result.cjs');
const sdkError = require('./core/sdk-error.cjs');
const ajaxError = require('./core/http/ajax-error.cjs');
const ajaxResult = require('./core/http/ajax-result.cjs');
const paramsFactory = require('./core/http/limiters/params-factory.cjs');
const rateLimiter = require('./core/http/limiters/rate-limiter.cjs');
const operatingLimiter = require('./core/http/limiters/operating-limiter.cjs');
const adaptiveDelayer = require('./core/http/limiters/adaptive-delayer.cjs');
const manager = require('./core/http/limiters/manager.cjs');
const versionManager = require('./core/version-manager.cjs');
const abstractB24 = require('./core/abstract-b24.cjs');
const v2 = require('./core/http/v2.cjs');
const v3 = require('./core/http/v3.cjs');
const useFormatters = require('./tools/use-formatters.cjs');
const environment = require('./tools/environment.cjs');
const filterV3 = require('./tools/filter-v3.cjs');
const batchRefV3 = require('./tools/batch-ref-v3.cjs');
const useB24Helper = require('./helper/use-b24-helper.cjs');
const loaderB24frame = require('./loader-b24frame.cjs');
const abstractLogger = require('./logger/abstract-logger.cjs');
const frame = require('./frame/frame.cjs');
const auth = require('./hook/auth.cjs');
const auth$1 = require('./frame/auth.cjs');
const auth$2 = require('./oauth/auth.cjs');
const b24$1 = require('./frame/b24.cjs');
const b24$2 = require('./hook/b24.cjs');
const b24$3 = require('./oauth/b24.cjs');
const client = require('./pullClient/client.cjs');
const consolaAdapter = require('./logger/handler/consola-adapter.cjs');
const consoleHandler = require('./logger/handler/console-handler.cjs');
const consoleV2Handler = require('./logger/handler/console-v2-handler.cjs');
const dialog = require('./frame/dialog.cjs');
const entityType = require('./types/crm/entity-type.cjs');
const jsonFormatter = require('./logger/formatter/json-formatter.cjs');
const lineFormatter = require('./logger/formatter/line-formatter.cjs');
const logger = require('./types/logger.cjs');
const logger$1 = require('./logger/logger.cjs');
const browser$1 = require('./logger/browser.cjs');
const loggerFactory = require('./logger/logger-factory.cjs');
const memoryHandler = require('./logger/handler/memory-handler.cjs');
const commands = require('./frame/message/commands.cjs');
const controller = require('./frame/message/controller.cjs');
const nullLogger = require('./logger/null-logger.cjs');
const options = require('./frame/options.cjs');
const parent = require('./frame/parent.cjs');
const placement = require('./frame/placement.cjs');
const productrow = require('./types/crm/productrow.cjs');
const refreshTokenError = require('./oauth/refresh-token-error.cjs');
const slider = require('./frame/slider.cjs');
const streamHandler = require('./logger/handler/stream-handler.cjs');
const telegramFormatter = require('./logger/formatter/telegram-formatter.cjs');
const telegramHandler = require('./logger/handler/telegram-handler.cjs');
const winstonAdapter = require('./logger/handler/winston-adapter.cjs');
const memoryUsageProcessor = require('./logger/processor/memory-usage-processor.cjs');
const pidProcessor = require('./logger/processor/pid-processor.cjs');



exports.DataType = common.DataType;
exports.Type = type.Type;
exports.getEnumValue = index$2.getEnumValue;
exports.isArrayOfArray = index$2.isArrayOfArray;
exports.omit = index$2.omit;
exports.pick = index$2.pick;
exports.Text = text.Text;
exports.Browser = browser.Browser;
exports.ApiVersion = b24.ApiVersion;
exports.CatalogProductImageType = index.CatalogProductImageType;
exports.CatalogProductType = index.CatalogProductType;
exports.CatalogRoundingRuleType = index.CatalogRoundingRuleType;
exports.EnumBitrix24Edition = index$1.EnumBitrix24Edition;
exports.EnumBizprocBaseType = index$1.EnumBizprocBaseType;
exports.EnumBizprocDocumentType = index$1.EnumBizprocDocumentType;
exports.convertBizprocDocumentTypeToCrmEntityTypeId = index$1.convertBizprocDocumentTypeToCrmEntityTypeId;
exports.getDocumentId = index$1.getDocumentId;
exports.getDocumentType = index$1.getDocumentType;
exports.getDocumentTypeForFilter = index$1.getDocumentTypeForFilter;
exports.EnumAppStatus = b24Helper.EnumAppStatus;
exports.LoadDataType = b24Helper.LoadDataType;
exports.StatusDescriptions = b24Helper.StatusDescriptions;
exports.TypeOption = b24Helper.TypeOption;
exports.TypeSpecificUrl = b24Helper.TypeSpecificUrl;
exports.CloseReasons = pull.CloseReasons;
exports.ConnectionType = pull.ConnectionType;
exports.ListRpcError = pull.ListRpcError;
exports.LsKeys = pull.LsKeys;
exports.PullStatus = pull.PullStatus;
exports.RpcMethod = pull.RpcMethod;
exports.SenderType = pull.SenderType;
exports.ServerMode = pull.ServerMode;
exports.SubscriptionType = pull.SubscriptionType;
exports.SystemCommands = pull.SystemCommands;
exports.B24LangList = list.B24LangList;
exports.B24LocaleMap = list.B24LocaleMap;
exports.Result = result.Result;
exports.SdkError = sdkError.SdkError;
exports.AjaxError = ajaxError.AjaxError;
exports.AjaxResult = ajaxResult.AjaxResult;
exports.ParamsFactory = paramsFactory.ParamsFactory;
exports.RateLimiter = rateLimiter.RateLimiter;
exports.OperatingLimiter = operatingLimiter.OperatingLimiter;
exports.AdaptiveDelayer = adaptiveDelayer.AdaptiveDelayer;
exports.RestrictionManager = manager.RestrictionManager;
exports.versionManager = versionManager.versionManager;
exports.AbstractB24 = abstractB24.AbstractB24;
exports.HttpV2 = v2.HttpV2;
exports.HttpV3 = v3.HttpV3;
exports.useFormatter = useFormatters.useFormatter;
exports.Environment = environment.Environment;
exports.getEnvironment = environment.getEnvironment;
exports.FilterV3 = filterV3.FilterV3;
exports.BatchRefV3 = batchRefV3.BatchRefV3;
exports.useB24Helper = useB24Helper.useB24Helper;
exports.initializeB24Frame = loaderB24frame.initializeB24Frame;
exports.AbstractLogger = abstractLogger.AbstractLogger;
exports.AppFrame = frame.AppFrame;
exports.AuthHookManager = auth.AuthHookManager;
exports.AuthManager = auth$1.AuthManager;
exports.AuthOAuthManager = auth$2.AuthOAuthManager;
exports.B24Frame = b24$1.B24Frame;
exports.B24Hook = b24$2.B24Hook;
exports.B24OAuth = b24$3.B24OAuth;
exports.B24PullClientManager = client.PullClient;
exports.ConsolaAdapter = consolaAdapter.ConsolaAdapter;
exports.ConsoleHandler = consoleHandler.ConsoleHandler;
exports.ConsoleV2Handler = consoleV2Handler.ConsoleV2Handler;
exports.DialogManager = dialog.DialogManager;
exports.EnumCrmEntityType = entityType.EnumCrmEntityType;
exports.EnumCrmEntityTypeId = entityType.EnumCrmEntityTypeId;
exports.EnumCrmEntityTypeShort = entityType.EnumCrmEntityTypeShort;
exports.getEnumCrmEntityTypeShort = entityType.getEnumCrmEntityTypeShort;
exports.JsonFormatter = jsonFormatter.JsonFormatter;
exports.LineFormatter = lineFormatter.LineFormatter;
exports.LogLevel = logger.LogLevel;
exports.Logger = logger$1.Logger;
exports.LoggerBrowser = browser$1.LoggerBrowser;
exports.LoggerType = browser$1.LoggerType;
exports.LoggerFactory = loggerFactory.LoggerFactory;
exports.MemoryHandler = memoryHandler.MemoryHandler;
exports.MessageCommands = commands.MessageCommands;
exports.MessageManager = controller.MessageManager;
exports.NullLogger = nullLogger.NullLogger;
exports.OptionsManager = options.OptionsManager;
exports.ParentManager = parent.ParentManager;
exports.PlacementManager = placement.PlacementManager;
exports.ProductRowDiscountTypeId = productrow.ProductRowDiscountTypeId;
exports.RefreshTokenError = refreshTokenError.RefreshTokenError;
exports.SliderManager = slider.SliderManager;
exports.StreamHandler = streamHandler.StreamHandler;
exports.TelegramFormatter = telegramFormatter.TelegramFormatter;
exports.TelegramHandler = telegramHandler.TelegramHandler;
exports.WinstonAdapter = winstonAdapter.WinstonAdapter;
exports.memoryUsageProcessor = memoryUsageProcessor.memoryUsageProcessor;
exports.pidProcessor = pidProcessor.pidProcessor;
