// @bitrix24/b24jssdk/dist/esm/frame/message/controller.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { Type } from '../../tools/type.mjs';
import { Text } from '../../tools/text.mjs';
import { omit } from '../../tools/index.mjs';
import { LoggerFactory } from '../../logger/logger-factory.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class MessageManager {
  static {
    __name(this, "MessageManager");
  }
  #appFrame;
  #callbackPromises;
  #callbackSingletone;
  // origins already warned about (#244) — dedup so a peer spamming postMessage
  // from a foreign origin can't flood a wired logger sink
  #rejectedOrigins = /* @__PURE__ */ new Set();
  _logger;
  runCallbackHandler;
  constructor(appFrame) {
    this._logger = LoggerFactory.createNullLogger();
    this.#appFrame = appFrame;
    this.#callbackPromises = /* @__PURE__ */ new Map();
    this.#callbackSingletone = /* @__PURE__ */ new Map();
    this.runCallbackHandler = this._runCallback.bind(this);
  }
  setLogger(logger) {
    this._logger = logger;
  }
  getLogger() {
    return this._logger;
  }
  // region Events ////
  /**
   * Subscribe to the onMessage event of the parent window
   */
  subscribe() {
    window.addEventListener("message", this.runCallbackHandler);
  }
  /**
   * Unsubscribe from the onMessage event of the parent window
   */
  unsubscribe() {
    window.removeEventListener("message", this.runCallbackHandler);
  }
  // endregion ////
  /**
   * Send message to parent window
   * The answer (if) we will get in _runCallback
   *
   * @param command
   * @param params
   */
  async send(command, params = null) {
    return new Promise((resolve, reject) => {
      let cmd;
      const promiseHandler = {
        resolve,
        reject,
        timeoutId: null
      };
      const keyPromise = this.#setCallbackPromise(promiseHandler);
      let paramsSend = null;
      const optionsSend = omit(params || {}, ["singleOption", "callBack", "isSafely", "safelyTime", "requestId"]);
      const { callBack, singleOption, requestId } = params || {};
      if (callBack) {
        this.#callbackSingletone.set(keyPromise, callBack);
      }
      if (singleOption) {
        paramsSend = singleOption;
      } else if (Object.keys(optionsSend).length > 0) {
        paramsSend = { ...optionsSend };
      }
      if (command.toString().includes(":")) {
        cmd = {
          method: command.toString(),
          params: paramsSend || "",
          callback: keyPromise,
          appSid: this.#appFrame.getAppSid(),
          requestId
        };
      } else {
        cmd = command.toString();
        if (params?.isRawValue !== true && paramsSend) {
          paramsSend = JSON.stringify(paramsSend);
        } else if (params?.isRawValue === true && paramsSend && Type.isPlainObject(paramsSend) && paramsSend["value"]) {
          paramsSend = paramsSend["value"];
        }
        const listParams = [
          paramsSend || "",
          keyPromise,
          this.#appFrame.getAppSid()
        ];
        cmd += ":" + listParams.filter(Boolean).join(":");
      }
      this.getLogger().debug(`send to ${this.#appFrame.getTargetOrigin()}`, {
        command: command.toString(),
        callbackKey: keyPromise,
        origin: this.#appFrame.getTargetOrigin()
      });
      parent.postMessage(cmd, this.#appFrame.getTargetOrigin());
      if (params?.isSafely) {
        const safelyTime = Number.parseInt(String(params?.safelyTime || 900));
        this.#callbackPromises.get(keyPromise).timeoutId = window.setTimeout(
          () => {
            if (this.#callbackPromises.has(keyPromise)) {
              this.getLogger().warning(`action ${command.toString()} stop by timeout`, {
                command: command.toString(),
                safelyTime
              });
              this.#callbackPromises.delete(keyPromise);
              resolve({ isSafely: true });
            }
          },
          safelyTime
        );
      }
    });
  }
  /**
   * Fulfilling a promise based on messages from the parent window
   *
   * @param event
   * @private
   */
  _runCallback(event) {
    if (event.origin !== this.#appFrame.getTargetOrigin()) {
      if (!this.#rejectedOrigins.has(event.origin)) {
        this.#rejectedOrigins.add(event.origin);
        this.getLogger().warning("message rejected: unexpected origin", {
          origin: event.origin,
          expected: this.#appFrame.getTargetOrigin()
        });
      }
      return;
    }
    if (event.data) {
      const tmp = event.data.split(":");
      const cmd = {
        id: tmp[0],
        args: tmp.slice(1).join(":")
      };
      this.getLogger().debug(`get from ${event.origin}`, {
        id: cmd.id,
        origin: event.origin
      });
      if (cmd.args) {
        cmd.args = JSON.parse(cmd.args);
      }
      if (this.#callbackPromises.has(cmd.id)) {
        const promise = this.#callbackPromises.get(cmd.id);
        if (promise.timeoutId) {
          clearTimeout(promise.timeoutId);
        }
        this.#callbackPromises.delete(cmd.id);
        promise.resolve(cmd.args);
      } else if (this.#callbackSingletone.has(cmd.id)) {
        const callBack = this.#callbackSingletone.get(cmd.id);
        if (callBack) {
          callBack.apply(globalThis, [cmd.args]);
        }
      }
    }
  }
  /**
   * Storing a promise for a message from the parent window
   *
   * @param promiseHandler
   * @private
   *
   * @memo We don't use Symbol here, because we need to pass it to the parent and then find and restore it.
   */
  #setCallbackPromise(promiseHandler) {
    const key = Text.getUniqId();
    this.#callbackPromises.set(key, promiseHandler);
    return key;
  }
}

export { MessageManager };
