// @bitrix24/b24jssdk/dist/cjs/types/bizproc/index.cjs
/**
 * @package @bitrix24/b24jssdk
 * @version 2.0.0
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
'use strict';

const entityType = require('../crm/entity-type.cjs');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var EnumBitrix24Edition = /* @__PURE__ */ ((EnumBitrix24Edition2) => {
  EnumBitrix24Edition2["undefined"] = "undefined";
  EnumBitrix24Edition2["b24"] = "b24";
  EnumBitrix24Edition2["box"] = "box";
  return EnumBitrix24Edition2;
})(EnumBitrix24Edition || {});
var EnumBizprocBaseType = /* @__PURE__ */ ((EnumBizprocBaseType2) => {
  EnumBizprocBaseType2["undefined"] = "undefined";
  EnumBizprocBaseType2["crm"] = "crm";
  EnumBizprocBaseType2["disk"] = "disk";
  EnumBizprocBaseType2["lists"] = "lists";
  return EnumBizprocBaseType2;
})(EnumBizprocBaseType || {});
var EnumBizprocDocumentType = /* @__PURE__ */ ((EnumBizprocDocumentType2) => {
  EnumBizprocDocumentType2["undefined"] = "undefined";
  EnumBizprocDocumentType2["lead"] = "CCrmDocumentLead";
  EnumBizprocDocumentType2["company"] = "CCrmDocumentCompany";
  EnumBizprocDocumentType2["contact"] = "CCrmDocumentContact";
  EnumBizprocDocumentType2["deal"] = "CCrmDocumentDeal";
  EnumBizprocDocumentType2["invoice"] = "Bitrix\\Crm\\Integration\\BizProc\\Document\\SmartInvoice";
  EnumBizprocDocumentType2["quote"] = "Bitrix\\Crm\\Integration\\BizProc\\Document\\Quote";
  EnumBizprocDocumentType2["order"] = "Bitrix\\Crm\\Integration\\BizProc\\Document\\Order";
  EnumBizprocDocumentType2["dynamic"] = "Bitrix\\Crm\\Integration\\BizProc\\Document\\Dynamic";
  EnumBizprocDocumentType2["disk"] = "Bitrix\\Disk\\BizProcDocument";
  EnumBizprocDocumentType2["lists"] = "BizprocDocument";
  EnumBizprocDocumentType2["listsList"] = "Bitrix\\Lists\\BizprocDocumentLists";
  return EnumBizprocDocumentType2;
})(EnumBizprocDocumentType || {});
function convertBizprocDocumentTypeToCrmEntityTypeId(documentType) {
  switch (documentType) {
    case "CCrmDocumentLead" /* lead */:
      return entityType.EnumCrmEntityTypeId.lead;
    case "CCrmDocumentCompany" /* company */:
      return entityType.EnumCrmEntityTypeId.company;
    case "CCrmDocumentContact" /* contact */:
      return entityType.EnumCrmEntityTypeId.contact;
    case "CCrmDocumentDeal" /* deal */:
      return entityType.EnumCrmEntityTypeId.deal;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\SmartInvoice" /* invoice */:
      return entityType.EnumCrmEntityTypeId.invoice;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Quote" /* quote */:
      return entityType.EnumCrmEntityTypeId.quote;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Order" /* order */:
      return entityType.EnumCrmEntityTypeId.order;
  }
  return entityType.EnumCrmEntityTypeId.undefined;
}
__name(convertBizprocDocumentTypeToCrmEntityTypeId, "convertBizprocDocumentTypeToCrmEntityTypeId");
function getDocumentType(documentType, entityId) {
  let entityIdFormatted = "";
  let base = "undefined" /* undefined */;
  switch (documentType) {
    case "CCrmDocumentLead" /* lead */:
      base = "crm" /* crm */;
      entityIdFormatted = "LEAD";
      break;
    case "CCrmDocumentCompany" /* company */:
      base = "crm" /* crm */;
      entityIdFormatted = "COMPANY";
      break;
    case "CCrmDocumentContact" /* contact */:
      base = "crm" /* crm */;
      entityIdFormatted = "CONTACT";
      break;
    case "CCrmDocumentDeal" /* deal */:
      base = "crm" /* crm */;
      entityIdFormatted = "DEAL";
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\SmartInvoice" /* invoice */:
      base = "crm" /* crm */;
      entityIdFormatted = "SMART_INVOICE";
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Quote" /* quote */:
      base = "crm" /* crm */;
      entityIdFormatted = "QUOTE";
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Order" /* order */:
      base = "crm" /* crm */;
      entityIdFormatted = "ORDER";
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Dynamic" /* dynamic */:
      base = "crm" /* crm */;
      entityIdFormatted = `DYNAMIC_${entityId || 0}`;
      if ((entityId || 0) < 1) {
        throw new Error("Need set entityId");
      }
      break;
    case "Bitrix\\Disk\\BizProcDocument" /* disk */:
      base = "disk" /* disk */;
      entityIdFormatted = `STORAGE_${entityId || 0}`;
      if ((entityId || 0) < 1) {
        throw new Error("Need set entityId");
      }
      break;
    case "BizprocDocument" /* lists */:
      base = "lists" /* lists */;
      entityIdFormatted = `iblock_${entityId || 0}`;
      if ((entityId || 0) < 1) {
        throw new Error("Need set entityId");
      }
      break;
    case "Bitrix\\Lists\\BizprocDocumentLists" /* listsList */:
      base = "lists" /* lists */;
      entityIdFormatted = `iblock_${entityId || 0}`;
      if ((entityId || 0) < 1) {
        throw new Error("Need set entityId");
      }
      break;
  }
  return [
    base,
    documentType,
    entityIdFormatted
  ];
}
__name(getDocumentType, "getDocumentType");
function getDocumentId(documentType, id, dynamicId) {
  let entityIdFormatted = "";
  const tmp = getDocumentType(documentType, 1);
  switch (documentType) {
    case "CCrmDocumentLead" /* lead */:
      entityIdFormatted = `LEAD_${id}`;
      break;
    case "CCrmDocumentCompany" /* company */:
      entityIdFormatted = `COMPANY_${id}`;
      break;
    case "CCrmDocumentContact" /* contact */:
      entityIdFormatted = `CONTACT_${id}`;
      break;
    case "CCrmDocumentDeal" /* deal */:
      entityIdFormatted = `DEAL_${id}`;
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\SmartInvoice" /* invoice */:
      entityIdFormatted = `SMART_INVOICE_${id}`;
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Quote" /* quote */:
      entityIdFormatted = `QUOTE_${id}`;
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Order" /* order */:
      entityIdFormatted = `ORDER_${id}`;
      break;
    case "Bitrix\\Crm\\Integration\\BizProc\\Document\\Dynamic" /* dynamic */:
      entityIdFormatted = `DYNAMIC_${dynamicId || 0}_${id}`;
      if ((dynamicId || 0) < 1) {
        throw new Error("Need set dynamicId");
      }
      break;
    case "Bitrix\\Disk\\BizProcDocument" /* disk */:
      entityIdFormatted = `${id}`;
      break;
    case "BizprocDocument" /* lists */:
      entityIdFormatted = `${id}`;
      break;
    case "Bitrix\\Lists\\BizprocDocumentLists" /* listsList */:
      entityIdFormatted = `${id}`;
      break;
  }
  return [
    tmp[0],
    tmp[1],
    entityIdFormatted
  ];
}
__name(getDocumentId, "getDocumentId");
function getDocumentTypeForFilter(documentType) {
  const result = getDocumentType(documentType, 1);
  return [
    result[0],
    result[1]
  ];
}
__name(getDocumentTypeForFilter, "getDocumentTypeForFilter");

exports.EnumBitrix24Edition = EnumBitrix24Edition;
exports.EnumBizprocBaseType = EnumBizprocBaseType;
exports.EnumBizprocDocumentType = EnumBizprocDocumentType;
exports.convertBizprocDocumentTypeToCrmEntityTypeId = convertBizprocDocumentTypeToCrmEntityTypeId;
exports.getDocumentId = getDocumentId;
exports.getDocumentType = getDocumentType;
exports.getDocumentTypeForFilter = getDocumentTypeForFilter;
